src/daemon.rs

```diff
@@ -0,0 +1,891 @@
+//    just fine for starters)
+//  - maybe pause / restart remote control
+//  - more flexible cadence computation
+//  - more test cases for the cadence computation
+
+// For documentation about configuration and message topics that are used, see ~/doc/HOWTO-DAEMON.md.
+
+// THREADS AND I/O
+//
+// The main thread of the daemon listens on a channel from which it reads events: alarms (for work
+// to do), signals (keyboard interrupts), and incoming messages (from some controlling agent).
+//
+// Signal handlers place signals in the daemon's channel as events.
+
+use crate::cluster;
+use crate::datasink::{DataSink, StdioSink};
+use crate::jobsapi;
+use crate::json_tags;
+use crate::ps;
+use crate::realsystem;
+use crate::slurmjobs;
+use crate::sysinfo;
+use crate::systemapi::SystemAPI;
+use crate::time::{unix_now, unix_time_components};
+
+use std::io::{BufRead, Write};
+use std::sync::mpsc;
+use std::thread;
+
+pub struct GlobalIni {
+    pub cluster: String,
+    pub role: String,
+    pub lockdir: Option<String>,
+}
+
+pub struct DebugIni {
+    pub dump: bool,
+    pub verbose: bool,
+}
+
+pub struct SampleIni {
+    pub cadence: Option<Dur>,
+    pub exclude_system_jobs: bool,
+    pub load: bool,
+    pub batchless: bool,
+    pub exclude_commands: Vec<String>,
+    pub exclude_users: Vec<String>,
+}
+
+pub struct SysinfoIni {
+    pub on_startup: bool,
+    pub cadence: Option<Dur>,
+}
+
+pub struct JobsIni {
+    pub cadence: Option<Dur>,
+    pub window: Option<Dur>,
+    pub uncompleted: bool,
+    pub delta_coding: bool,
+    pub dump: Option<String>,
+}
+
+pub struct ClusterIni {
+    pub cadence: Option<Dur>,
+}
+
+pub struct Ini {
+    pub global: GlobalIni,
+    pub debug: DebugIni,
+    pub sample: SampleIni,
+    pub sysinfo: SysinfoIni,
+    pub jobs: JobsIni,
+    pub cluster: ClusterIni,
+}
+
+#[derive(Clone, Debug)]
+pub enum Operation {
+    Sample,
+    Sysinfo,
+    Jobs,
+    Cluster,
+    // Signals: signal code
+    #[allow(dead_code)]
+    Signal(i32),
+    // Control messages: key/value
+    #[allow(dead_code)]
+    Incoming(String, String),
+    // Unrecoverable errors on other threads: explanatory message
+    #[allow(dead_code)]
+    Fatal(String),
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum Dur {
+    Hours(u64),
+    Minutes(u64),
+    Seconds(u64),
+}
+
+#[allow(dead_code)]
+impl Dur {
+    pub fn to_seconds(self) -> u64 {
+        match self {
+            Dur::Hours(n) => n * 60 * 60,
+            Dur::Minutes(n) => n * 60,
+            Dur::Seconds(n) => n,
+        }
+    }
+
+    pub fn to_minutes(self) -> u64 {
+        match self {
+            Dur::Hours(n) => n * 60,
+            Dur::Minutes(n) => n,
+            Dur::Seconds(n) => (n + 59) / 60,
+        }
+    }
+}
+
+// The daemon_mode() should return under these circumstances *only*:
+//
+// - it fails to parse the ini file
+// - it fails to read the api token or password files
```

Comment:
"... any specified files" is probably more appropriate, also merge with next line